airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py

```diff
@@ -0,0 +1,128 @@
+SOFTWARE.
+"""
+
+from apiclient import errors
+from typing import Generator
+
+from airbyte_protocol import AirbyteConnectionStatus, Status
+from airbyte_protocol import AirbyteCatalog
+from airbyte_protocol import AirbyteLogger
+from airbyte_protocol import AirbyteMessage, Type
+from airbyte_protocol import Source
+
+from .helpers import Helpers
+from .models.generated.spreadsheet_values import *
+from .models.generated.spreadsheet import *
+
+ROW_BATCH_SIZE = 200
+
+
+class GoogleSheetsSource(Source):
+    """
+    Spreadsheets API Reference: https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def check(self, logger: AirbyteLogger, config_container) -> AirbyteConnectionStatus:
+        # Check involves verifying that the specified spreadsheet is reachable with our credentials.
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            # Attempt to get first row of sheet
+            client.get(spreadsheetId=spreadsheet_id, includeGridData=False, ranges='1:1').execute()
+        except errors.HttpError as err:
+            reason = str(err)
+            # Give a clearer message if it's a common error like 404.
+            if err.resp.status == 404:
+                reason = "Requested spreadsheet was not found."
+
+            print(f"Formatted error: {reason}")
+            return AirbyteConnectionStatus(status=Status.FAILED, message=str(reason))
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def discover(self, logger: AirbyteLogger, config_container) -> AirbyteCatalog:
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            logger.info(f"Running discovery on sheet {spreadsheet_id}")
+            spreadsheet_metadata = Spreadsheet.parse_obj(client.get(spreadsheetId=spreadsheet_id, includeGridData=False).execute())
+            sheet_names = [sheet.properties.title for sheet in spreadsheet_metadata.sheets]
+            streams = []
+            for sheet_name in sheet_names:
+                header_row_data = Helpers.get_first_row(client, spreadsheet_id, sheet_name)
+                stream = Helpers.headers_to_airbyte_stream(sheet_name, header_row_data)
+                streams.append(stream)
+            return AirbyteCatalog(streams=streams)
+
+        except errors.HttpError as err:
+            reason = str(err)
+            if err.resp.status == 404:
+                reason = "Requested spreadsheet was not found."
+            raise Exception(f"Could not run discovery: {reason}")
+
+    def read(self, logger: AirbyteLogger, config_container, catalog_path, state=None) -> Generator[AirbyteMessage, None, None]:
```

Comment:
same as discover. is there a unit test that could be written for this logic (or some subset of it). alot of transformations going on.